pkg/tfbridge/validate_input_types.go

```diff
@@ -0,0 +1,471 @@
+		propObjectValue := propertyValue.ObjectValue()
+		stableKeys := propObjectValue.StableKeys()
+
+		// handle required properties
+		if len(requiredObjectProps) > 0 {
+			if len(stableKeys) == 0 {
+				failures = append(failures, TypeFailure{
+					ResourcePath: pathBuilder.toPath(),
+					Reason: fmt.Sprintf(
+						"%s object is missing required properties: %s",
+						propertyKey, strings.Join(requiredObjectProps, ", "),
+					),
+				})
+				return &failures
+			}
+			for _, requiredProp := range requiredObjectProps {
+				if !slices.Contains(stableKeys, resource.PropertyKey(requiredProp)) {
+					failures = append(failures, TypeFailure{
+						ResourcePath: pathBuilder.toPath(),
+						Reason: fmt.Sprintf(
+							"%s object is missing required property: %s",
+							propertyKey, requiredProp,
+						),
+					})
+				}
+			}
+			if len(failures) > 0 {
+				return &failures
+			}
+		}
+
+		for _, objectKey := range stableKeys {
+			objectValue := propObjectValue[objectKey]
+			if complexSpec != nil {
+				complexProps, ok := complexSpec.Properties[string(objectKey)]
+				if !ok {
+					failures = append(failures, TypeFailure{
+						ResourcePath: pathBuilder.toPath(),
+						Reason: fmt.Sprintf(
+							"property %s is not defined in the schema",
+							string(objectKey),
+						),
+					})
+					continue
+				}
+				objectTypeSpec = &complexProps.TypeSpec
+			}
+			if objectTypeSpec != nil {
+				if failure := v.validateTypeSpec(
+					string(objectKey),
+					objectValue,
+					*objectTypeSpec,
+					pathBuilder,
+					requiredObjectProps,
+				); failure != nil {
+					failures = append(failures, *failure...)
+				}
+			}
+		}
+	// The number and string cases are for enums. The normal cases are handled
+	// in the matchType function right now we are not validating enums. Enums
+	// are more of a helper, but are not exhaustive in all cases i.e. ec2
+	// instance types
+	case "string":
+	case "number":
+	}
+
+	if len(failures) > 0 {
+		return &failures
+	}
+
+	return nil
+}
+
+// getType gets a type definition from a schema reference. Currently it only
+// supports types from the same schema
+func (v *PulumiInputValidator) getType(typeRef string) *pschema.ComplexTypeSpec {
+	if strings.HasPrefix(typeRef, "#/types/") {
+		ref := strings.TrimPrefix(typeRef, "#/types/")
+		if typeSpec, ok := v.schema.Types[ref]; ok {
+			return &typeSpec
+		}
+	}
+	return nil
+}
+
+// ValidateInputs will validate a set of inputs against the pulumi schema. It will
+// return a list of type failures if any are found
+func (v *PulumiInputValidator) ValidateInputs(inputs resource.PropertyMap) *[]TypeFailure {
+	failures := []TypeFailure{}
+	for key, value := range inputs {
+		if failure := v.validateInputType(string(key), value, pathBuilder{paths: []string{}}); failure != nil {
+			failures = append(failures, *failure...)
+		}
+	}
+
+	if len(failures) > 0 {
+		return &failures
+	}
+
+	return nil
+}
+
+func (v *PulumiInputValidator) validateInputType(
```

Comment:
Does this validate resources only or Invoke (what TF calls DataSource) inputs as well? I suspect only the former, in which case it would be great to highlight that with a better name and comment.